File: packages/platform_android_tts/android/src/main/kotlin/com/example/platform_android_tts/services/KokoroTtsService.kt
package com.example.platform_android_tts.services

import android.app.Service
import android.content.Intent
import android.os.IBinder
import com.example.platform_android_tts.sherpa.KokoroSherpaInference
import kotlinx.coroutines.*
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Kokoro TTS Service running in its own process (:kokoro).
 * 
 * Provides synthesis using Kokoro sherpa-onnx TTS models.
 * Kokoro is a multi-speaker model - all voices share a single model
 * with different speaker IDs.
 * 
 * Uses KokoroSherpaInference for actual TTS synthesis (via sherpa-onnx).
 */
class KokoroTtsService : Service() {
    
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    
    // Synthesis counter to prevent unload during active synthesis
    private val synthesisCounter = SynthesisCounter()
    
    // Model state - Kokoro uses single shared model
    private var modelPath: String? = null
    private var voicesPath: String? = null
    private var isInitialized = false
    private var loadedVoices = mutableMapOf<String, KokoroVoice>() // voiceId -> voice info
    
    // Single inference engine (shared model)
    private var inference: KokoroSherpaInference? = null
    
    // Active synthesis jobs for cancellation
    private val activeJobs = mutableMapOf<String, Job>()
    
    override fun onBind(intent: Intent?): IBinder? = null
    
    override fun onDestroy() {
        scope.cancel()
        unloadAllModels()
        super.onDestroy()
    }
    
    /**
     * Initialize the Kokoro engine with the core model path.
     * 
     * @param corePath Path to directory containing model.onnx, tokens.txt, espeak-ng-data/
     * @param voicesFile Path to voices.bin file with voice embeddings
     */
    suspend fun initEngine(corePath: String, voicesFile: String? = null): Result<Unit> = runCatching {
        if (isInitialized && modelPath == corePath) {
            return Result.success(Unit)
        }
        
        // Verify model files exist - support both model.onnx and model.int8.onnx
        val modelFile = findModelFile(corePath)
        if (modelFile == null || !modelFile.exists()) {
            throw IllegalStateException("Model file not found in: $corePath")
        }
        
        // Look for voices file in expected locations
        val actualVoicesPath = voicesFile ?: findVoicesFile(corePath)
        if (actualVoicesPath == null || !File(actualVoicesPath).exists()) {
            throw IllegalStateException("Voices file not found in $corePath")
        }
        
        // Initialize sherpa-onnx inference engine
        val engine = KokoroSherpaInference(corePath, actualVoicesPath)
        engine.initialize().getOrThrow()
        
        inference = engine
        modelPath = corePath
        voicesPath = actualVoicesPath
        isInitialized = true
    }
    
    /**
     * Find the ONNX model file in the core path.
     */
    private fun findModelFile(corePath: String): File? {
        val standardModel = File(corePath, "model.onnx")
        if (standardModel.exists()) return standardModel
        
        val int8Model = File(corePath, "model.int8.onnx")
        if (int8Model.exists()) return int8Model
        
        // Look for any .onnx file
        return File(corePath).listFiles { file ->
            file.isFile && file.name.endsWith(".onnx")
        }?.maxByOrNull { it.length() }
    }
    
    /**
     * Find the voices.bin file in expected locations.
     */
    private fun findVoicesFile(corePath: String): String? {
        val candidates = listOf(
            "$corePath/voices.bin",
            "$corePath/../voices.bin",
            "$corePath/voices/voices.bin"
        )
        return candidates.firstOrNull { File(it).exists() }
    }
    
    /**
     * Load a specific voice (register speaker ID mapping).
     */
    suspend fun loadVoice(voiceId: String, speakerId: Int): Result<Unit> = runCatching {
        if (!isInitialized) {
            throw IllegalStateException("Engine not initialized")
        }
        
        loadedVoices[voiceId] = KokoroVoice(
            voiceId = voiceId,
            speakerId = speakerId,
            lastUsed = System.currentTimeMillis()
        )
    }
    
    /**
     * Synthesize text to a WAV file.
     * 
     * Note: Kokoro uses 24000 Hz sample rate.
     */
    suspend fun synthesize(
        voiceId: String,
        text: String,
        outputPath: String,
        requestId: String,
        speakerId: Int = 0,
        speed: Float = 1.0f
    ): SynthesisResult {
        val engine = inference
        
        if (!isInitialized || engine == null) {
            return SynthesisResult(
                success = false,
                errorCode = ErrorCode.MODEL_MISSING,
                errorMessage = "Engine not initialized"
            )
        }
        
        if (text.isBlank()) {
            return SynthesisResult(
                success = false,
                errorCode = ErrorCode.INVALID_INPUT,
                errorMessage = "Empty text"
            )
        }
        
        // Track active synthesis to prevent unload during operation
        synthesisCounter.increment()
        
        // Get speaker ID from loaded voice or use provided ID
        val voice = loadedVoices[voiceId]
        val actualSpeakerId = voice?.speakerId ?: speakerId
        
        val job = scope.launch {
            try {
                val tmpFile = File("$outputPath.tmp")
                tmpFile.parentFile?.mkdirs()
                
                // Run sherpa-onnx inference (only once!)
                val samples = engine.synthesize(text, actualSpeakerId, speed).getOrThrow()
                val sampleRate = engine.getSampleRate()
                
                // Convert float samples to 16-bit PCM
                val pcmData = samples.map { sample ->
                    (sample * 32767).toInt().coerceIn(-32768, 32767).toShort()
                }.toShortArray()
                
                writeWavFile(tmpFile, pcmData, sampleRate)
                
                ensureActive()
                tmpFile.renameTo(File(outputPath))
                
            } catch (e: CancellationException) {
                File("$outputPath.tmp").delete()
                throw e
            }
        }
        
        activeJobs[requestId] = job
        
        return try {
            job.join()
            
            if (job.isCancelled) {
                SynthesisResult(
                    success = false,
                    errorCode = ErrorCode.CANCELLED,
                    errorMessage = "Synthesis cancelled"
                )
            } else {
                voice?.lastUsed = System.currentTimeMillis()
                
                // Calculate duration from output file size (avoids double synthesis)
                val sampleRate = engine.getSampleRate()
                val outputFile = File(outputPath)
                val durationMs = if (outputFile.exists()) {
                    val fileSize = outputFile.length()
                    val dataSize = fileSize - 44 // WAV header is 44 bytes
                    val sampleCount = dataSize / 2 // 16-bit samples = 2 bytes each
                    (sampleCount * 1000 / sampleRate).toInt()
                } else {
                    estimateDurationMs(text)
                }
                
                SynthesisResult(
                    success = true,
                    durationMs = durationMs,
                    sampleRate = sampleRate
                )
            }
        } catch (e: Exception) {
            SynthesisResult(
                success = false,
                errorCode = mapExceptionToErrorCode(e),
                errorMessage = e.message
            )
        } finally {
            activeJobs.remove(requestId)
            synthesisCounter.decrement()
        }
    }
    
    /**
     * Cancel an in-flight synthesis.
     */
    fun cancelSynthesis(requestId: String) {
        activeJobs[requestId]?.cancel()
        activeJobs.remove(requestId)
    }
    
    /**
     * Unload a specific voice (just removes from tracking).
     * Waits for any active synthesis to complete first.
     */
    fun unloadVoice(voiceId: String) {
        // Wait for any active synthesis to complete (max 5 seconds)
        if (!synthesisCounter.waitUntilIdle(timeoutMs = 5000)) {
            android.util.Log.w("KokoroTtsService", "Timeout waiting for synthesis to complete before unload")
        }
        loadedVoices.remove(voiceId)
    }
    
    /**
     * Unload all models and reset state.
     * Waits for any active synthesis to complete first.
     */
    fun unloadAllModels() {
        // Wait for any active synthesis to complete (max 5 seconds)
        if (!synthesisCounter.waitUntilIdle(timeoutMs = 5000)) {
            android.util.Log.w("KokoroTtsService", "Timeout waiting for synthesis to complete before unloadAll")
        }
        
        inference?.dispose()
        inference = null
        loadedVoices.clear()
        isInitialized = false
        modelPath = null
        voicesPath = null
        
        activeJobs.values.forEach { it.cancel() }
        activeJobs.clear()
    }
    
    /**
     * Unload least recently used voice.
     */
    fun unloadLeastUsedVoice(): String? {
        val lru = loadedVoices.minByOrNull { it.value.lastUsed }
        return lru?.let {
            loadedVoices.remove(it.key)
            it.key
        }
    }
    
    /**
     * Get memory info.
     */
    fun getMemoryInfo(): ServiceMemoryInfo {
        val runtime = Runtime.getRuntime()
        val availableMB = ((runtime.maxMemory() - runtime.totalMemory()) / (1024 * 1024)).toInt()
        val totalMB = (runtime.maxMemory() / (1024 * 1024)).toInt()
        
        return ServiceMemoryInfo(
            availableMB = availableMB,
            totalMB = totalMB,
            loadedModelCount = if (isInitialized) 1 else 0
        )
    }
    
    /**
     * Check if engine is ready.
     */
    fun isReady(): Boolean = isInitialized && inference?.isReady() == true
    
    /**
     * Check if a voice is loaded.
     */
    fun isVoiceLoaded(voiceId: String): Boolean = loadedVoices.containsKey(voiceId)
    
    // Private helpers
    
    private fun estimateDuration(text: String): Float {
        // Rough estimate: 150 words per minute, average 5 characters per word
        val words = text.length / 5.0f
        return words / 2.5f // 150 wpm = 2.5 words per second
    }
    
    private fun estimateDurationMs(text: String): Int {
        return (estimateDuration(text) * 1000).toInt()
    }
    
    private fun writeWavFile(file: File, samples: ShortArray, sampleRate: Int) {
        val numChannels = 1
        val bitsPerSample = 16
        val byteRate = sampleRate * numChannels * bitsPerSample / 8
        val blockAlign = numChannels * bitsPerSample / 8
        val dataSize = samples.size * 2
        val fileSize = 36 + dataSize
        
        RandomAccessFile(file, "rw").use { raf ->
            // RIFF header
            raf.writeBytes("RIFF")
            raf.writeIntLE(fileSize)
            raf.writeBytes("WAVE")
            
            // fmt chunk
            raf.writeBytes("fmt ")
            raf.writeIntLE(16) // chunk size
            raf.writeShortLE(1) // audio format (PCM)
            raf.writeShortLE(numChannels)
            raf.writeIntLE(sampleRate)
            raf.writeIntLE(byteRate)
            raf.writeShortLE(blockAlign)
            raf.writeShortLE(bitsPerSample)
            
            // data chunk
            raf.writeBytes("data")
            raf.writeIntLE(dataSize)
            
            // Write samples
            val buffer = ByteBuffer.allocate(samples.size * 2).order(ByteOrder.LITTLE_ENDIAN)
            samples.forEach { buffer.putShort(it) }
            raf.write(buffer.array())
        }
    }
    
    private fun RandomAccessFile.writeIntLE(value: Int) {
        val buffer = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN)
        buffer.putInt(value)
        write(buffer.array())
    }
    
    private fun RandomAccessFile.writeShortLE(value: Int) {
        val buffer = ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN)
        buffer.putShort(value.toShort())
        write(buffer.array())
    }
    
    private fun mapExceptionToErrorCode(e: Exception): ErrorCode {
        return when {
            e is OutOfMemoryError -> ErrorCode.OUT_OF_MEMORY
            e is CancellationException -> ErrorCode.CANCELLED
            e.message?.contains("model", ignoreCase = true) == true -> ErrorCode.MODEL_MISSING
            else -> ErrorCode.INFERENCE_FAILED
        }
    }
}

/**
 * Result of a synthesis operation.
 */
data class SynthesisResult(
    val success: Boolean,
    val durationMs: Int? = null,
    val sampleRate: Int? = null,
    val errorCode: ErrorCode? = null,
    val errorMessage: String? = null
)

/**
 * Memory information (service-level type).
 */
data class ServiceMemoryInfo(
    val availableMB: Int,
    val totalMB: Int,
    val loadedModelCount: Int
)

/**
 * Error codes for synthesis operations.
 */
enum class ErrorCode {
    NONE,
    MODEL_MISSING,
    MODEL_CORRUPTED,
    OUT_OF_MEMORY,
    INFERENCE_FAILED,
    CANCELLED,
    RUNTIME_CRASH,
    INVALID_INPUT,
    FILE_WRITE_ERROR,
    UNKNOWN
}

/**
 * Kokoro voice state.
 */
data class KokoroVoice(
    val voiceId: String,
    val speakerId: Int,
    var lastUsed: Long
)

File: packages/platform_ios_tts/ios/Classes/engines/KokoroTtsService.swift
import Foundation
import SherpaOnnxCApi

/// Kokoro TTS engine service.
/// Uses sherpa-onnx's native Kokoro TTS support for inference.
class KokoroTtsService: TtsServiceProtocol {
    let engineType: NativeEngineType = .kokoro
    
    private var inference = KokoroSherpaInference()
    private var loadedVoices: [String: VoiceInfo] = [:]
    private let lock = NSLock()
    private let synthesisCounter = SynthesisCounter()
    
    var isReady: Bool {
        lock.lock()
        defer { lock.unlock() }
        return inference.isModelLoaded
    }
    
    func loadCore(corePath: String, configPath: String?) async throws {
        // For Kokoro, the core model is loaded with voice
        lock.lock()
        defer { lock.unlock() }
        
        guard FileManager.default.fileExists(atPath: corePath) else {
            throw TtsError.modelNotLoaded
        }
        
        print("[KokoroTtsService] Core path validated: \(corePath)")
    }
    
    func loadVoice(voiceId: String, modelPath: String, speakerId: Int?, configPath: String?) async throws {
        lock.lock()
        defer { lock.unlock() }
        
        // modelPath is the directory containing the model files
        let modelDir = modelPath
        
        guard FileManager.default.fileExists(atPath: modelDir) else {
            throw TtsError.voiceNotLoaded(voiceId)
        }
        
        // Find the ONNX model file in the directory
        let fm = FileManager.default
        guard let contents = try? fm.contentsOfDirectory(atPath: modelDir) else {
            throw TtsError.invalidInput("Cannot read model directory: \(modelDir)")
        }
        
        // Find .onnx file (not .onnx.json)
        let onnxFile = contents.first { $0.hasSuffix(".onnx") && !$0.hasSuffix(".onnx.json") }
        guard let onnxFile = onnxFile else {
            throw TtsError.invalidInput("No ONNX model found in: \(modelDir)")
        }
        
        let fullModelPath = (modelDir as NSString).appendingPathComponent(onnxFile)
        
        // Kokoro requires: model.onnx, voices.bin, tokens.txt
        let voicesPath = (modelDir as NSString).appendingPathComponent("voices.bin")
        let tokensPath = (modelDir as NSString).appendingPathComponent("tokens.txt")
        
        guard FileManager.default.fileExists(atPath: voicesPath) else {
            throw TtsError.invalidInput("Voices file not found at: \(voicesPath)")
        }
        
        guard FileManager.default.fileExists(atPath: tokensPath) else {
            throw TtsError.invalidInput("Tokens file not found at: \(tokensPath)")
        }
        
        // Find espeak-ng-data directory if exists
        let dataDir = (modelDir as NSString).appendingPathComponent("espeak-ng-data")
        let dataDirPath = FileManager.default.fileExists(atPath: dataDir) ? dataDir : nil
        
        print("[KokoroTtsService] Loading model from: \(fullModelPath)")
        print("[KokoroTtsService] Voices file: \(voicesPath)")
        print("[KokoroTtsService] Tokens file: \(tokensPath)")
        print("[KokoroTtsService] Data dir: \(dataDirPath ?? "none")")
        
        // Load the model
        try inference.loadModel(
            modelPath: fullModelPath,
            voicesPath: voicesPath,
            tokensPath: tokensPath,
            dataDir: dataDirPath
        )
        
        loadedVoices[voiceId] = VoiceInfo(
            voiceId: voiceId,
            modelPath: fullModelPath,
            speakerId: speakerId,
            lastUsed: Date()
        )
        print("[KokoroTtsService] Voice loaded: \(voiceId)")
    }
    
    func synthesize(text: String, voiceId: String, outputPath: String, speakerId: Int?, speed: Double) async throws -> SynthesizeResult {
        // Track active synthesis to prevent unload during operation
        synthesisCounter.increment()
        defer { synthesisCounter.decrement() }
        
        lock.lock()
        let voice = loadedVoices[voiceId]
        lock.unlock()
        
        guard voice != nil else {
            throw TtsError.voiceNotLoaded(voiceId)
        }
        
        guard !text.isEmpty else {
            throw TtsError.invalidInput("Text is empty")
        }
        
        // Synthesize using sherpa-onnx
        let (samples, sampleRate) = try inference.synthesize(
            text: text,
            speakerId: speakerId ?? 0,
            speed: Float(speed)
        )
        
        // Write WAV file
        try AudioConverter.writeWav(samples: samples, sampleRate: sampleRate, to: outputPath)
        
        let durationMs = AudioConverter.durationMs(sampleCount: samples.count, sampleRate: sampleRate)
        
        return SynthesizeResult(
            success: true,
            durationMs: durationMs,
            sampleRate: Int64(sampleRate),
            errorCode: .none,
            errorMessage: nil
        )
    }
    
    func cancelSynthesis(requestId: String) {
        print("[KokoroTtsService] Cancel requested for: \(requestId)")
    }
    
    func unloadVoice(voiceId: String) {
        // Wait for active synthesis operations to complete
        _ = synthesisCounter.waitUntilIdle(timeoutMs: 5000)
        
        lock.lock()
        defer { lock.unlock() }
        loadedVoices.removeValue(forKey: voiceId)
        inference.unload()
        print("[KokoroTtsService] Voice unloaded: \(voiceId)")
    }
    
    func unloadAll() {
        // Wait for active synthesis operations to complete
        _ = synthesisCounter.waitUntilIdle(timeoutMs: 5000)
        
        lock.lock()
        defer { lock.unlock() }
        loadedVoices.removeAll()
        inference.unload()
        print("[KokoroTtsService] All resources unloaded")
    }
}

/// Voice information stored in memory.
struct VoiceInfo {
    let voiceId: String
    let modelPath: String
    let speakerId: Int?
    var lastUsed: Date
}
